### ISSUE TITLE
Deferred type annotations

### ISSUE THREAD

**Issue by hkan** [decahedronio/entity#7]
Currently entities cannot have circular type annotations _(Blog annotating Comment, and Comment annotation Blog back)_ because JavaScript cannot handle circular dependencies. If I'm not missing any other possibility, one can only have circular types through custom `fromJson` and `toJson` implementations, which does not look very pretty to me and I try to avoid it as far as I can. 😅 

**Side note**: Circular dependency is when two JavaScript modules (files) import each other. ([More on circular dependencies](https://spin.atomicobject.com/2018/06/25/circular-dependencies-javascript/))

[Objection.js](https://github.com/Vincit/objection.js/) (a popular JavaScript ORM) solves this issue by resolving model modules only when the model is needed, instead of resolving it on app initialization. (See "Require loops" under "Relations" in Objection.js' documentation.)

My proposal is the same logic with a different format. Let's resolve this issue by creating a second `@Type` annotation as `@TypeDeferred` (or `@TypeLazy` or anything else, I'm not the best on naming). It'll simply append a flag to `TypeMetadata` and `TypeMetadata` will resolve deferred types when `TypeMetada.type` is requested. A proxy object would probably work best here. Example code is at the end.

I fiddled around with this idea in my local environment and it seems to work fine but I wanted to check-in with at least another person to actually spending time on a PR.

```typescript
// Proposed TypeMetadata structure
export class TypeMetadata {
    constructor(public target: Function,
                public propertyName: string,
                public sourcePropertyName: string,
                private _type: Function,
                public deferred: boolean = false) {
    }

    public get type() {
        return this.deferred ? this._type() : this._type;
    }
}

// TypeDeferred annotator function
export function TypeDeferred(type: () => Promise<Function>, jsonKey?: string) {
    return function (target: any, key: string) {
        jsonKey = jsonKey ? jsonKey : (
            EntityBuilder.enableCamelConversion ? StringHelper.toSnake(key) : key
        );

        const metadata = new TypeMetadata(target.constructor, key, jsonKey, type, true);
        defaultMetadataStorage.addTypeMetadata(metadata);
    };
}

// Example model
export class BlogPost extends Entity {
    @TypeDeferred(() => import('./Comment'))
    public comments: Comment = undefined;
}
```

**Comment by phroggyy** [comment]
Hey @hkan – this is a great idea. However, I'm wondering if you think it'd make most sense to make a new `@TypeDeferred`, or implicitly defer it if users pass a function? e.g
```
@Type(() => import('./Comment'))
```

Something that might be neat could be something like a double decorator, a la
```
@Defer
@Type()
```
But I don't think we can get to any significantly nicer syntax.

My ideal would be
```
@Defer
@Type(Comment)
```
but that obviously won't work as it'll cause `Comment` to be immediately resolved. I'm open to proposals though. Not a massive fan of passing a resolver function into the decorator, but I'm also not sure we have many alternatives.

In summary: I'm in favour of this. I'd prefer to not pass a resolver function, but don't know of a way to get around that.

**Comment by hkan** [comment]
My initial thought was to pass a `() => import('./Comment')` to the decorator but since both options are just a `Function` in JavaScript's perspective, I don't know how we'd differentiate between a resolver and an entity constructor. Double decorator can be a workaround on this, and I agree that it looks a little nicer. One other option would be to pass an object like `{ entity: () => import('./Comment') }` but that's just worse syntax in my opinion.

Another way to do this is with strings as a path to entity file. That way EntityBuilder can resolve the class from the path. Two catches here: 1st one is relative paths will be fragile, and 2nd is the referenced path should export the entity as default.

    @Type('@/entities/Comment')

**Comment by phroggyy** [comment]
I had a think about it, and indeed I too thought about a string initially, but I think it's _better_ (although not quite as sexy) to have a closure – most IDEs will know how to properly refactor an `import()` statement, while they're much less likely to understand how to refactor a string.

Further, using a string would mean entity starts relying on `import` being available; I'd much rather have that be a user-land decision. So I think let's go ahead with the original solution. I still think it would be neat if we could figure out a way to distinguish between `() => {}` and `Comment`. Obviously tough since they're both `Function`, but I feel there should be a way to do it...

If there isn't, let's just do `@DeferredType`

**Comment by hkan** [comment]
Alright, I've found it! [Arrow functions do not have `prototype`s](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#Use_of_prototype_property). We can go with the arrow function resolver, without introducing new decorators.

```typescript
const resolver = () => import('./Foo');

resolver.prototype === undefined; // true
```

There are however 2 other things now.

- Which class should know about if a given type annotation is a resolver function or an entity constructor? (and obviously resolve it if encounters a resolver function.) I'm thinking `TypeMetadata` would be suitable since it's the class that already knows about which attribute gets which type.
- Since `import()` function brings in the async stuff, this feature will be a breaking change because async effect bubbles up all the way to `EntityBuilder`. Building entities will require an `await` or a `.then`.

**Comment by phroggyy** [comment]
hm, is it possible to move that to the arrow function? e.g `() => await import('./Foo')`? just thinking if there are alternatives for making it non-breaking

**Comment by hkan** [comment]
It is possible but it would still affect the _consumer end_. `await` returns a `Promise` and there is no way to avoid waiting for that promise to resolve with a simple approach. One can only _delegate_ the waiting via resolving all `() => import(...)` types after all the model files are imported on app start, but that requires an additional file like the one below:

```typescript
import './Blog';
import './Comment';
import './User';

const metadatas: TypeMetadata[] = getAllDefinedTypeMetadata();

metadatas.forEach(async (metadata) => {
  // If type is an arrow function, we will assume it returns an 'import'.
  if (metadata.type.prototype === undefined) {
    // Run the function to actually import the module and assign the module to type prop
    // so that the EntityBuilder will actually get an entity constructor, and not a Promise.
    metadata.type = await metadata.type();
  }
});
```

This will push users to update their codebase to not actually start until all the imports are resolved. Otherwise a race condition or a downright error might happen.

**Comment by hkan** [comment]
I think I've found a way, but I need to implement all the functionality, and thoroughly test it to be sure. Apparently, in Webpack's output of `require()` is simply the sync version of `import()` which is super nice, because with `require()` we can implement this feature with backwards compat. I'll get on that, see how it goes.

**Comment by phroggyy** [comment]
hm – right. I'm _hoping_ to keep this as much on the user end as possible, as to impose minimum requirements for users.
